### PR TITLE
path: dynapath delta 40-60 post processor add

### PR DESCRIPTION
**Pull request by luvtofish**
Initial upload

Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

**Comment by luvtofish** [comment]
Link to Path forum discussion on this post processor. 
https://forum.freecadweb.org/viewtopic.php?f=15&t=74207

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [715037095](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/715037095) was triggered at [8b9af20](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/8b9af206abfa6f436d433cf698c1858a1fed7e61). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [715160482](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/715160482) was triggered at [51f43db](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/51f43db5e1d56b7dd4de98ba7e7bb7feee3c9b1f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [715173545](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/715173545) was triggered at [e0ad54d](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/e0ad54decef4bf6830a96d22d6a55e465c9aae49). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [715367498](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/715367498) was triggered at [60b5d81](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/60b5d819b6713e556759d086b10a4490d68c71c7). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [716130729](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/716130729) was triggered at [f611cc6](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/f611cc66b45a3a90db47bd7adc67b09b4401e1b9). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [716155978](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/716155978) was triggered at [efce625](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/efce625a8d4252e3bdf66534ed90fdd887e0973b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [717225661](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/717225661) was triggered at [0d8aaca](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/0d8aacae49b61b5d16d685033abcd80debd37145). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [717249470](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/717249470) was triggered at [3ac6bc8](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/3ac6bc8ea6cb1388d01a0c7131c82982bb443494). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [717259337](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/717259337) was triggered at [c7bbac5](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/c7bbac501ed290a21f6947a3fb36bd16303895f4). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [719236940](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/719236940) was triggered at [21c4b2d](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/21c4b2dcce8e1ecde7d3d96f514e9f88d0bf4ee0). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [719237659](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/719237659) was triggered at [db01062](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/db01062242b9031cd82f24f0dfeec789f4fbcb06). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [722911145](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/722911145) was triggered at [d8b6474](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/d8b6474e6d4ecb5faaf1be96a6db4a0a620a871a). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Path/Path/Post/scripts/dynapath_4060_post.py [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+                )
+            gcode += linenumber() + "M9" + "\n"
+
+    # do the post_amble
+    if OUTPUT_COMMENTS:
+        gcode += linenumber() + "(T)" + "BEGIN POSTAMBLE$\n"
+    for line in POSTAMBLE.splitlines(True):
+        gcode += linenumber() + line
+    gcode += "E\n"
+
```
Comment (LarryWoestman) [inline src/Mod/Path/Path/Post/scripts/dynapath_4060_post.py:329]: Should there be an "E" in that line?

**Comment by luvtofish** [comment]
On the Dynapath Control, the letter E signifies the end of program. Without
that character at the end, the control throws a loading error when loading
a program from external media. It’s otherwise ignored.

File: src/Mod/Path/Path/Post/scripts/dynapath_4060_post.py (hunk 1 above)
Comment (sliptonic) [reply src/Mod/Path/Path/Post/scripts/dynapath_4060_post.py:329]: If the E is an End of program, why not include it in the postamble?

**Comment by sliptonic** [comment]
I don't have a problem accepting this PR.  The author hasn't done anything wrong.  The problem is that our postprocessor structure doesn't do any kind of inheritance so every new postprocessor duplicates tons of code.  The real problem is that we can't look at a postprocessor like this and see what makes a dynapath40-60 unique and different from the default or linuxcnc posts.  

Things like the trailing 'E' that are unique to dynapath should absolutely be commented like crazy.

**Comment by luvtofish** [comment]
That E gets stripped from the program upon loading into the control. Im not
able to load programs into the control without it.

moving to the postamble is a nice tip. I’ll try that now and see what
happens. I’ll comment after on the results.

**Comment by luvtofish** [comment]
It did what I expected it might, it's treating this as a command so it's placing a line number in front of it. This will bomb on the control because it will assume that line is part of the program and will bomb with "No EOF file found" upon loading in the machine.
N0379(T)BEGIN POSTAMBLE$
N0380M05
N0381G80
N0382G40
N0383G17
N0384G90
N0385M30
N0386E

**Comment by LarryWoestman** [comment]
I agree with sliptonic that this PR is acceptable according to the "old style" postprocessors.

I am working on refactoring the postprocessors so that the "what makes this postprocessor" unique is easier to see and to remove a lot of the code duplication.  That is still a work in progress with more to do.

Comments in the code that describe things that are unique to the dynapath can be very helpful for anyone who comes along later who is trying to understand this code and what a dynapath needs in its gcode.  Based on my personal experience, that can also include the original author after a year or more :-)

In particular it might save someone a fair amount of time by documenting what works and doesn't work for handling the "E" that we have been talking about.

**Comment by luvtofish** [comment]
> I agree with sliptonic that this PR is acceptable according to the "old style" postprocessors.
> 
> I am working on refactoring the postprocessors so that the "what makes this postprocessor" unique is easier to see and to remove a lot of the code duplication. That is still a work in progress with more to do.
> 
> Comments in the code that describe things that are unique to the dynapath can be very helpful for anyone who comes along later who is trying to understand this code and what a dynapath needs in its gcode. Based on my personal experience, that can also include the original author after a year or more :-)
> 
> In particular it might save someone a fair amount of time by documenting what works and doesn't work for handling the "E" that we have been talking about.

Refactoring the the code to eliminate the need for duplication will be incredibly helpful. As it stands today, I've had to repeat much of the code from linuxcnc but have also had to massage the data further to meet the needs of my controller. The G codes and requisite parameters just are not in sync so it requires a little pushing, squeezing, and stretching to get things to work.

I'll add more comments for the unique dynapath items and push up an update and yes, I need this as much for myself as for any poor sop who comes along later and attempts to decipher things. LOL 

There were quite a few things I had to modify including:

-  fixture codes
-  convert incremental arcs to absolute
-  line numbers that are 4 digits long and start with an N
-  no spaces between parameters
-  E - eof terminator for code upload to control
-  parenthesis around the program name at the beginning of the g-code
-  parenthesis around comments including insertion of the character "T" which stands for text field
-  inserting X0Y0 to go with the Z movement on first move after tool change (breaks without)
-  conversion of and insertion of new parameters "O" for clearance height, "L" for dwell, and others.
-  several other things I'm not thinking of at the moment but were equally important.

BTW, to get clean output using this PP, I've had to modify a couple python Path files and recompile. The incessant use of "G0 "Z + safeHeight" seems a bit over used. It will run fine with out the modification, it's just redundant and not needed. Also, after the G80 call it throws another one which I also comment out in the python source. 

Thanks Larry and Sliptonic!

**Comment by luvtofish** [comment]
Okay -- I've added multiple comments to the pp in and effort to point out the "unique" needs of the dynapath controller. Let me know if there are any other questions. Thanks gents!

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [730275871](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/730275871) was triggered at [660f635](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/660f63557bad4ce0811c23e9a11a33c701ac16a3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by LarryWoestman** [comment]
The CI checks are failing with misspelled words and whitespace at the ends of a couple of lines.

Other than that it looks reasonable to me.

**Comment by luvtofish** [comment]
I’ll check it and update. Thanks

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962">pipeline status</a> for feature branch [PR_7962](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7962). Pipeline [731211386](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/731211386) was triggered at [0289e3a](https://github.com/FreeCAD/FreeCAD/pull/7962/commits/0289e3ae52984a6266f1edc1d4d15287f2befecb). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
What's left here ?

**Comment by luvtofish** [comment]
It’s ready to merge if @LarryWoestman agrees.

**Comment by LarryWoestman** [comment]
I agree that it is ready to merge.  Is there something else that I need to do to approve it?

**Review by LarryWoestman — APPROVED** [review]

**Comment by LarryWoestman** [comment]
I think I found something that might be the way to approve it.

**Comment by luvtofish** [comment]
@LarryWoestman should just need to click the “files changed” tab and select the review drop down on the right.

**Pull request merged**